op/k8s/block_device_link_update.go

```diff
@@ -0,0 +1,115 @@
+package k8s
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cybozu-go/cke"
+	"github.com/cybozu-go/cke/op"
+	"github.com/cybozu-go/log"
+	"github.com/cybozu-go/well"
+)
+
+type blockDeviceLinkUpdateOp struct {
+	apiServer *cke.Node
+	nodes     []*cke.Node
+	step      int
+}
+
+// BlockDeviceLinkUpdateOp returns an Operator to restart kubelet
+func BlockDeviceLinkUpdateOp(apiServer *cke.Node, nodes []*cke.Node) cke.Operator {
+	return &blockDeviceLinkUpdateOp{apiServer: apiServer, nodes: nodes}
+}
+
+func (o *blockDeviceLinkUpdateOp) Name() string {
+	return "block-device-link-update"
+}
+
+func (o *blockDeviceLinkUpdateOp) Targets() []string {
+	ips := make([]string, len(o.nodes))
+	for i, n := range o.nodes {
+		ips[i] = n.Address
+	}
+	return ips
+}
+
+func (o *blockDeviceLinkUpdateOp) NextCommand() cke.Commander {
+	switch o.step {
+	case 0:
+		o.step++
+		return updateBlockDeviceLinkForV1_17(o.apiServer, o.nodes)
+	default:
+		return nil
+	}
+}
+
+type updateBlockDeviceLinkForV1_17Command struct {
+	apiServer *cke.Node
+	nodes     []*cke.Node
+}
+
+// updateBlockDeviceLinkForV1_17 move raw block device files.
+// This command is used for upgrading to k8s 1.17
+func updateBlockDeviceLinkForV1_17(apiServer *cke.Node, nodes []*cke.Node) cke.Commander {
+	return updateBlockDeviceLinkForV1_17Command{apiServer: apiServer, nodes: nodes}
+}
+
+func (c updateBlockDeviceLinkForV1_17Command) Run(ctx context.Context, inf cke.Infrastructure, _ string) error {
+	begin := time.Now()
+	env := well.NewEnvironment(ctx)
+	for _, n := range c.nodes {
+		n := n
+		env.Go(func(ctx context.Context) error {
+			clientset, err := inf.K8sClient(ctx, c.apiServer)
+			if err != nil {
+				return err
+			}
+
+			agent := inf.Agent(n.Address)
+			if agent == nil {
+				return errors.New("unable to prepare agent for " + n.Nodename())
+			}
+
+			stdout, stderr, err := agent.Run(fmt.Sprintf("sudo find %s -maxdepth 1 -name pvc-* -type d", op.CSIBlockDevicePublishDirectory))
```

Comment:
CKE does not require the user of SSH can become `root`.
Instead, CKE requires the user to be able to execute `docker run`.

When `root` privilege is needed, CKE should run scripts inside `cke-tools` container.